internal/server/data/providergroup.go

```diff
@@ -0,0 +1,249 @@
+	if err != nil {
+		return nil, handleReadError(err)
+	}
+	pg := (*models.ProviderGroup)(providerGroup)
+
+	members, err := listProviderGroupMembers(tx, pg)
+	if err != nil {
+		return nil, fmt.Errorf("retrieve provider group members: %w", err)
+	}
+	pg.Members = members
+
+	return pg, nil
+}
+
+func listProviderGroupMembers(tx ReadTxn, pg *models.ProviderGroup) ([]models.ProviderUser, error) {
+	stmt := `
+		SELECT pu.identity_id, pu.provider_id, pu.email, pu.last_update, pu.redirect_url, pu.access_token, pu.refresh_token, pu.expires_at FROM provider_users AS pu
+		JOIN provider_groups_provider_users ON provider_groups_provider_users.provider_user_identity_id = pu.identity_id AND provider_groups_provider_users.provider_id = pu.provider_id
+		WHERE pu.provider_id = ? AND provider_group_name = ?;
+	`
+	rows, err := tx.Query(stmt, pg.ProviderID, pg.Name)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	result := []models.ProviderUser{}
+	for rows.Next() {
+		var pu models.ProviderUser
+
+		// TODO: replace fields with (*providerUserTable)(&providerUser).ScanFields()
+		if err := rows.Scan(&pu.IdentityID, &pu.ProviderID, &pu.Email, &pu.LastUpdate, &pu.RedirectURL, &pu.AccessToken, &pu.RefreshToken, &pu.ExpiresAt); err != nil {
+			return nil, err
+		}
+		result = append(result, pu)
+	}
+	return result, rows.Err()
+}
+
+type ListProviderGroupOptions struct {
+	ByProviderID       uid.ID
+	ByMemberIdentityID uid.ID
+}
+
+// ListProviderGroups returns all provider groups that match the specified criteria
+func ListProviderGroups(tx ReadTxn, opts ListProviderGroupOptions) ([]models.ProviderGroup, error) {
+	table := &providerGroupTable{}
+	query := querybuilder.New("SELECT")
+	query.B(columnsForSelect(table))
+	query.B("FROM")
+	query.B(table.Table())
+
+	if opts.ByMemberIdentityID != 0 {
+		query.B(`
+			JOIN provider_groups_provider_users 
+			ON provider_groups.name = provider_groups_provider_users.provider_group_name 
+			AND provider_groups.provider_id = provider_groups_provider_users.provider_id
+			`)
+	}
+
+	query.B("WHERE 1=1") // this is a condition which is always true, it allows us to append more conditions if needed
+	if opts.ByProviderID != 0 {
+		query.B("AND provider_groups.provider_id = ?", opts.ByProviderID)
+	}
+	if opts.ByMemberIdentityID != 0 {
+		query.B(`AND provider_groups_provider_users.provider_user_identity_id = ?`, opts.ByMemberIdentityID)
+	}
+
+	rows, err := tx.Query(query.String(), query.Args...)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var result []models.ProviderGroup
+	for rows.Next() {
+		var providerGroup models.ProviderGroup
+		fields := (*providerGroupTable)(&providerGroup).ScanFields()
+
+		if err := rows.Scan(fields...); err != nil {
+			return nil, err
+		}
+
+		result = append(result, providerGroup)
+	}
+
+	// load group members
+	for i := range result {
+		members, err := listProviderGroupMembers(tx, &result[i])
```

Comment:
this is going to be really really slow; is it really required? should be optional if so and ideally we should join it into the same query.

Reply:
Removed in the new changes